### DISCUSSION TITLE
Dropping w64devkit support in favor of MSYS2

### DISCUSSION THREAD

**Discussion by mtmcp** [ggml-org/llama.cpp#8229] [General]
After working on integrating Vulkan with ggml, the minimalistic nature of w64devkit is starting to cause some problems. It is forcing special-case changes to `Makefile` only because it does not support package management like other Unix varieties, forcing developers to provide custom solutions to already-solved problems.

Every other Unix variant (including MSYS2) can support standard packages provided by `pkg-config`. Since we can utilize MSYS2 and simplify management of `Makefile` instead of being held back by deficiencies in `w64devkit`, it may be optimal to abandon it in favor of MSYS2 because maintaining multiple `mingw-w64` environments on Windows is redundant, and we should probably choose the best option in addition to native Visual Studio support.

Another possibility would be at very least to expect that `w64devkit` _must_ load third party dependencies in using `pkg-config`. As the [authors state](https://github.com/skeeto/w64devkit?tab=readme-ov-file#library-installation), it has facilities for installing packages in this way. That would require folks to maintain things like `vulkan.pc` files for loading Vulkan, and other libraries which may not build by default in GGML. But, in my experience, this will likely amount to duplicating packages from MSYS2 (like shaderc, and Vulkan headers, etc).

**Comment by ggerganov** [comment]
I don't have a strong preference. I've had a little bit of experience with MSYS2 in the past and it worked pretty well for my needs, so I would be ok with this change.

**Comment by mtmcp** [comment]
What I'm reading on the [w64devkit docs](https://github.com/skeeto/w64devkit?tab=readme-ov-file#library-installation) is that `pkg-config` .pc files can be placed alongside dependencies in there. In particular:

> 1. Install it under the sysroot at w64devkit/$ARCH/. The easiest option, but will require re-installation after upgrading w64devkit. If it defines .pc files, the pkg-config command will automatically find and use them.

So w64devkit users simply need to manually copy libraries/includes into `w64devkit/x86_64-w64-mingw32` and also add `w64devkit/x86_64-w64-mingw32/lib/pkgconfig` folder containing, for example, `vulkan.pc` file, which references the library as such:

```
prefix=/w64devkit/x86_64-w64-mingw32
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include

Name: Vulkan-Loader
Description: Vulkan Loader
Version: 1.3.283
Libs: -L${libdir} -lvulkan-1.lib
Cflags: -I${includedir}
```
In order to find template pkg-config files, in my MSYS2 environment I ran `cat $(pkg-config --path vulkan) > w64devkit/x86_64-w64-mingw32/vulkan.pc` and simply modify anything that differs.

By mandating to build this way, we can ensure overall Unix compatibility and we will then be able to remove any special case logic to the Makefiles. Due to the nature of how w64devkit is designed, this is the correct approach and any package management tasks are deferred to the users.
